### PR TITLE
add post and delete to visibility settings

### PR DESCRIPTION
**Pull request by gustavomm19**
https://github.com/4GeeksAcademy/About-4Geeks-Academy/issues/4421

### CODE REVIEW

**Review by tommygonzaleza — APPROVED** [review]

File: breathecode/events/views.py [hunk 1]
```diff
@@ -725,6 +725,58 @@ def get(self, request, event_type_slug, academy_id=None):
...
+        if not item:
+            raise ValidationException('Event type visibility setting not found',
+                                      404,
+                                      slug='event-type-visibility-setting-not-found')
+
+        event_type = EventType.objects.filter(slug=event_type_slug).first()
+        if not event_type:
+            raise ValidationException('Event type not found', slug='event-type-not-found')
+
+        other_event_type = EventType.objects.filter(visibility_settings__id=visibility_setting_id).exclude(
```
Comment (jefer94) [inline breathecode/events/views.py:768]: Delete this visibility just in this `EventType`, a academy is not able to edit the visibility setting of another academy, if no one `EventType` of any academy is using this visibilitysetting, this must be removed

File: breathecode/events/views.py [hunk 2]
```diff
@@ -725,6 +725,58 @@ def get(self, request, event_type_slug, academy_id=None):
...
+
+        syllabus = None
+        if 'syllabus' in request.data:
+            syllabus = Syllabus.objects.filter(id=request.data['syllabus']).first()
+
+        cohort = None
+        if 'cohort' in request.data:
+            cohort = Cohort.objects.filter(id=request.data['cohort']).first()
+
+        visibility_setting, created = EventTypeVisibilitySetting.objects.get_or_create(syllabus=syllabus,
```
Comment (jefer94) [inline breathecode/events/views.py:748]: Then, you got the instance of visibility setting, so you need add this to the many to many to `EventType` if it ain't already in it

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/events/views.py [hunk 3]
```diff
@@ -725,6 +725,58 @@ def get(self, request, event_type_slug, academy_id=None):
...
+        event_type = EventType.objects.filter(slug=event_type_slug).first()
+        if not event_type:
+            raise ValidationException('Event type not found', slug='event-type-not-found')
+
+        academy = Academy.objects.filter(id=request.data['academy']).first()
+        if not academy:
+            raise ValidationException('Academy not found', slug='academy-not-found')
+
+        syllabus = None
+        if 'syllabus' in request.data:
```
[the remark this answers is not in the document]
Comment (jefer94) [reply breathecode/events/views.py:677]: This is a example of how to do this https://github.com/breatheco-de/apiv2/blob/f8023078ea23dd457a6fcf954d64348fca530488/breathecode/admissions/views.py#L1372

File: breathecode/events/views.py (hunk 2 above)
Comment (jefer94) [reply breathecode/events/views.py:688]:
```py
if visibility_setting not in event_type.visibility_settings:
    event_type.visibility_settings.add(visibility_setting)
```

File: breathecode/events/views.py [hunk 4]
```diff
@@ -663,6 +663,61 @@ def get(self, request, event_type_slug, academy_id=None):
...
+        if not event_type:
+            raise ValidationException('Event type not found', slug='event-type-not-found')
+
+        if not item:
+            raise ValidationException('Event type visibility setting not found',
+                                      404,
+                                      slug='event-type-visibility-setting-not-found')
+
+        other_event_type = EventType.objects.filter(visibility_settings__id=visibility_setting_id,
+                                                    academy=academy_id).exclude(slug=event_type_slug).first()
```
Comment (jefer94) [inline breathecode/events/views.py:711]: `.exists()`

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            role=1,
+            capability='read_event_type',
+        )
+
+        response = self.client.get(url)
+        json = response.json()
+        expected = {'detail': 'not-found', 'status_code': 400}
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 400)
```
Comment (jefer94) [inline breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py:66]: Add a case when you is trying of get a visibility settings and it is found

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting_id.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            event_type_visibility_setting=True,
+            event_type={
+                'slug': 'funny_event',
+                'icon_url': 'https://www.google.com'
+            },
+        )
+
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, 204)
```
Comment (jefer94) [inline breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting_id.py:124]: Add database test with
`self.bc.database.list_of('events.EventTypeVisibilitySetting')`

Comment (jefer94) [reply breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting_id.py:124]: Add a test to check the ids
```py
self.bc.check.queryset_with_pks(model.event_type.visibility_settings.all(), [2])
```

Where 2 is the primary key of resource added to many to many and `model.event_type.visibility_settings.all()` is the many to many to be checked

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting_id.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            capability='crud_event_type',
+            event_type_visibility_setting=True,
+        )
+
+        response = self.client.delete(url)
+        json = response.json()
+        expected = {'detail': 'event-type-not-found', 'status_code': 400}
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 400)
```
Comment (jefer94) [inline breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting_id.py:100]: Add the test of database to the rest of tests

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            role=1,
+            capability='crud_event_type',
+        )
+
+        response = self.client.post(url)
+        json = response.json()
+        expected = {'detail': 'event-type-not-found', 'status_code': 400}
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 400)
```
Comment (jefer94) [inline breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py:110]: Add test of database to the rest of tests

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            'syllabus': {
+                'id': model.syllabus.id,
+                'name': model.syllabus.name,
+                'slug': model.syllabus.slug,
+            }
+        }
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 201)
+        self.bc.check.queryset_with_pks(model.event_type.visibility_settings.all(), [1])
```
Comment (jefer94) [inline breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py:205]: Check also the properties is being put in `EventTypeVisibilitySetting`

File: breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py (hunk 8 above)
Comment (jefer94) [reply breathecode/events/tests/urls/tests_academy_eventype_slug_visibilitysetting.py:111]: So think you is modifying 2 tables, `EventTypeVisibilitySetting` and the handler between it and `EventType`

**Review by jefer94 — CHANGES_REQUESTED** [review]

**Review by jefer94 — APPROVED** [review]

**Pull request merged**